### ISSUE TITLE
Docker 1.13 overwrite/remove named volume when stack deploy/update 

### ISSUE THREAD

**Issue by BastienAr** [moby/moby#31095]
Currently there is no option when updating a stack (`docker stack deploy`) to overwrite a named volume created by this stack. In a same way, there is no option (such -v) to remove named volume associated with a stack when you remove the stack, other than removing the volumes with `docker volume rm`.

Here is a usecase :

Given this file:
```yaml
version: '3'

services:
  config:
    image: test/config-hoster:1.3
    volumes:
      - volume_config:/datas
    deploy:
      mode: global
      restart_policy:
        condition: none
  nginx:
    image: nginx
    volumes:
      - volume_config:/datas:ro
    depends_on:
      - config


volumes:
  volume_config:
```

where *config-hoster* is a from scratch image, exposing a volume with the config files in it, and *nginx* a service used just for testing the correct mapping.

When I first deploy this stack, everything is fine, the volume is created **and** filled with the config files of the *config* service. But if I change the version (tag) of the config service, when I update the stack the volume is not deleted, so datas inside are obviously not overwrited. Same behavior if I rm and redeploy the stack.

Since this is a named volume, not directly bounded to an host directory, and not created externaly, I see no problem to erase or overwrite this volume when the stack is removed or updated.

**Comment by cpuguy83** [comment]
Hmm, this would actually be a pretty big problem if implemented.
The reason it's doing this is the volume exists already and is populated with data when you update the config.

I think the right thing here is to actually support config objects (like we do secrets).

**Comment by BastienAr** [comment]
I've just checked the secret objects in docker. This is great, and actually solve a problem we were facing with ssh keys. Indeed supporting config objects in a same way will perfecly match the previous usecase.

I'm not sure to understand why and what would be the problem of deleting volumes created inside a stack, if you're sure of what you're doing (hence the *-v* option). Could you clarify this for me please ?

**Comment by cpuguy83** [comment]
@BastienAr Volumes are meant for persistent storage. If you are telling a service to use a named volume, this is making a statement on the retention of this volume.
For instance if you use an anonymous volume (don't specify a name), this would be cleaned up.

**Comment by zh99998** [comment]
I'm using NFS as volume storage backend
and.... I can't find any way to change its options except stop containers, remove volumes and deploy again.

**Comment by BastienAr** [comment]
@cpuguy83 Ok, I get the philosophy. So we definitively need a way to start a service requiring a file that could be modified (like config files) on any swarm node. Currently if you have to change configuration, either you rebuild the image with correct config (dirty) or you bind volume and become host dependent. NFS is also a solution, but I find it cumbersome for this purpose.

**Comment by cpuguy83** [comment]
@BastienAr I feel like volumes are not a good use for injecting configuration, even though it may be the only way to do it right now.

One option that is currently available is to inject the config as a secret. It's a little hacky but works well in the short-term.

**Comment by BastienAr** [comment]
Yeah, that could be a trick, but it's still inconvenient due to the fixed path of secrets (/run/secret/<secret_name>). And the encryption of the file could add unnecessary process (config files rarely need encryption)

**Comment by thaJeztah** [comment]
> And the encryption of the file could add unnecessary process (config files rarely need encryption)

I think that overhead can be neglected; the encryption/decryption is part of the raft store, and should not cause a noticeable performance issue.

w.r.t. the paths (defaulting to `/run/secret/<name>`), the `<name>` part can be overridden when using the secret (`--secret src=my-little-secret,target=config.json`), but the _path_ inside the container currently not (we kept this possibility open for a future enhancement, but needs additional discussion). You _can_ create a symlink inside the container if needed.

**Comment by BastienAr** [comment]
Okay, this solution is quite simple and operational. Still I think this could be cleaner if shared file that are not secret would be separated from the term _secret_, even if under the hood this is the exact same process.

Bty is there any chance we could talk about this subject at the next Dockercon ? Some guys from our company (including me) will be there.

**Comment by BastienAr** [comment]
resolved in 17.06 by the introduction of [Docker configs](https://docs.docker.com/engine/swarm/configs/). I close the issue
